### ISSUE TITLE
[Bug] Failure to acquire gradient with requires_grad and backward(gradient) in cuda (with 0.4.0)

### ISSUE THREAD

**Issue by ryonakamura** [pytorch/pytorch#8287]
The following snippet is a reproduction of bug.
```python
import torch
import torch.nn as nn

x = torch.randn(2, 2, requires_grad=True).cuda(0)
t = torch.randn(2, 2).cuda(0)
loss_func = nn.MSELoss().cuda(0)
l1 = nn.Sequential(*[nn.Linear(2, 2) for _ in range(3)]).cuda(0)
l2 = nn.Sequential(*[nn.Linear(2, 2) for _ in range(3)]).cuda(0)
h1 = l1(x)
h2 = h1.detach().requires_grad_()
h3 = l2(h2)
l = loss_func(h3, t)
l.backward()
print('h2:', h2.grad)
h1.backward(gradient=h2.grad)
print('x:', x.grad)
```
It is possible to acquire gradient of `x` in cpu, but it fails in cuda.

**Comment by zou3519** [comment]
> x = torch.randn(2, 2, requires_grad=True).cuda(0)

If you want gradients to accumulate in `x`, you should change the above line to 
``` 
x = torch.randn(2, 2, requires_grad=True, device='cuda:0')
```

The reason why your code doesn't work is because of 
> x = torch.randn(2, 2, requires_grad=True).cuda(0)

What this is doing is creating a "leaf" tensor `(torch.randn(2, 2, requires_grad=True)` and then assigning a copy of it (that is on CUDA device 0) to `x`. Gradients can only accumulate in leaf tensors.
